### PR TITLE
CRM: Changing location of externally added import Transaction menu links

### PR DESCRIPTION
**Pull request by coder-karen**
## Proposed changes:

* This PR changes the location of the 'Import' link that is added when the CSV Pro extension is installed and active, under the 'Transactions' menu. Previously it was located at the bottom of the menu underneath a separator. Now it is located above the 'Export' link, to be more consistent with the placement in the Contact menu.
* If any extra menu items are added to the 'Transactions' menu via the `zbs-transactions-menu` filter, they will be added to the bottom of the menu underneath a separator.
* Another option was to make the menu / filter extensible in a way that would allow being able to add any new item at a specified location, but I've not taken that route (initially) for reasons mentioned here: https://github.com/Automattic/zero-bs-crm/issues/2896#issuecomment-1494617655

### Other information:

- [ ] Have you written new tests for your changes, if applicable?
- [ ] Have you checked the E2E test CI results, and verified that your changes do not break them?
- [ ] Have you tested your changes on WordPress.com, if applicable (if so, you'll see a generated comment below with a script to run)?

## Jetpack product discussion

https://github.com/Automattic/zero-bs-crm/issues/2896

## Does this pull request change what data or activity we track or use?

No.

## Testing instructions:

* To test this out, on a local development site without this branch checked out (or a Jurassic Ninja site with either the Jetpack CRM current version or even the Jetpack Beta Tester plugin active and set to bleeding edge or latest stable), make sure to add the CSV Pro extension and activate it. 
* The default behaviour when looking at the Transactions menu dropdown (from any CRM dashboard page) is that the 'Import' link will be at the bottom of the menu, underneath a separator.
* Checkout this branch locally or via the Beta Tester plugin and check the Transactions menu again - the 'Import' link should now be above the 'Export' link.

**Before:**
Screenshot 2023-04-04 at 10 40 08

**After:**

Screenshot 2023-04-04 at 10 38 27

### CODE REVIEW

File: projects/plugins/crm/includes/ZeroBSCRM.Core.Menus.Top.php [hunk 1]
```diff
@@ -565,25 +565,35 @@ function zeroBSCRM_admin_top_menu( $branding = 'zero-bs-crm', $page = 'dash' ) {
...
 					<a class="item" href="<?php echo jpcrm_esc_link( 'tags', -1, 'zerobs_transaction', false, 'zerobscrm_transactiontag' ); ?>"><i class="icon tags"></i> <?php esc_html_e( 'Tags', 'zero-bs-crm' ); ?></a>
 
-					<?php if ( zeroBSCRM_permsTransactions() ) { ?>
+					<?php
+					// If CSV Pro is installed and active it will add an Import menu item to the zbs-transactions-menu filter - we'll then add that here
+					$transactions_menu = array();
+					$transactions_menu = apply_filters( 'zbs-transactions-menu', $transactions_menu ); // phpcs:ignore WordPress.NamingConventions.ValidHookName.UseUnderscores
+					if ( count( $transactions_menu ) > 0 ) {
+						foreach ( $transactions_menu as $menu_item ) {
+							if ( preg_match( '/\bimport\b/i', $menu_item ) ) {
```
Comment (cleacos) [inline projects/plugins/crm/includes/ZeroBSCRM.Core.Menus.Top.php]: If the menu is translated, it wouldn't work

Comment (coder-karen) [reply projects/plugins/crm/includes/ZeroBSCRM.Core.Menus.Top.php]: Oh good catch! I'll come up with an alternative.

Comment (coder-karen) [reply projects/plugins/crm/includes/ZeroBSCRM.Core.Menus.Top.php]: I added a commit which checks for the existence of the URL parameter that leads to the import page instead.

File: projects/plugins/crm/includes/ZeroBSCRM.Core.Menus.Top.php [hunk 2]
```diff
@@ -565,25 +565,35 @@ function zeroBSCRM_admin_top_menu( $branding = 'zero-bs-crm', $page = 'dash' ) {
...
+					$transactions_menu = array();
+					$transactions_menu = apply_filters( 'zbs-transactions-menu', $transactions_menu ); // phpcs:ignore WordPress.NamingConventions.ValidHookName.UseUnderscores
+					if ( count( $transactions_menu ) > 0 ) {
+						foreach ( $transactions_menu as $menu_item ) {
+							if ( preg_match( '/\bpage\=zerobscrm\-csvimporter\-app\b/i', $menu_item ) ) {
+								echo wp_kses( $menu_item, $zbs->acceptable_html );
+								$transactions_menu = array_diff( $transactions_menu, array( $menu_item ) );
+							}
+						}
+					}
```
Comment (cleacos) [inline projects/plugins/crm/includes/ZeroBSCRM.Core.Menus.Top.php]: Why is this block outside of `zeroBSCRM_permsTransactions` condition?

Comment (cleacos) [reply projects/plugins/crm/includes/ZeroBSCRM.Core.Menus.Top.php]: Also, maybe we could use `preg_grep` instead of the `foreach`

Comment (cleacos) [reply projects/plugins/crm/includes/ZeroBSCRM.Core.Menus.Top.php]: and even without the `count( $transaction_menu ) > 0` condition

File: projects/plugins/crm/includes/ZeroBSCRM.Core.Menus.Top.php [hunk 3]
```diff
@@ -565,25 +565,29 @@ function zeroBSCRM_admin_top_menu( $branding = 'zero-bs-crm', $page = 'dash' ) {
...
 
-					<?php if ( zeroBSCRM_permsTransactions() ) { ?>
+					<?php
+					if ( zeroBSCRM_permsTransactions() ) {
+						// If CSV Pro is installed and active it will add an Import menu item to the zbs-transactions-menu filter - we'll then add that here
+						$transactions_menu = array();
+						$transactions_menu = apply_filters( 'zbs-transactions-menu', $transactions_menu ); // phpcs:ignore WordPress.NamingConventions.ValidHookName.UseUnderscores
+						$import_menu_item  = preg_grep( '/\bpage\=zerobscrm\-csvimporter\-app\b/i', $transactions_menu );
+						echo wp_kses( $import_menu_item[0], $zbs->acceptable_html );
+						$transactions_menu = array_diff( $transactions_menu, $import_menu_item );
```
Comment (cleacos) [inline projects/plugins/crm/includes/ZeroBSCRM.Core.Menus.Top.php]: Here, we have to check if we get the menu or the array is empty (count = 0)

Comment (cleacos) [reply projects/plugins/crm/includes/ZeroBSCRM.Core.Menus.Top.php]: If I deactivate the CSV Importer Pro extension I get the notice:

image

**Review by cleacos — APPROVED** [review]
Great! Now it works perfectly! LGTM!

**Pull request merged**